### PR TITLE
fix: preserve TradingView historical range data(OK-57783, OK-57784)

### PR DESCRIPTION
**Pull request by limichange**
[OK-57783](https://onekeyhq.atlassian.net/browse/OK-57783) [OK-57784](https://onekeyhq.atlassian.net/browse/OK-57784)

----------

## Summary
- Split non-native market K-line requests into backend-safe 200-point slices with one-interval overlap.
- Merge, sort, deduplicate, and clip history data to `[from, to)` so historical date ranges remain continuous.
- Add regression coverage for backend response caps and exclusive slice boundaries.

## Intent & Context
OK-57783 and OK-57784 cover incorrect native TradingView date navigation and missing historical K-line data. Selecting a historical range could land on the requested window while the candles inside that window were absent, leaving a flat-looking fragment or a blank middle section.

This is the app-side half of the coordinated date-range fix. Chart companion: https://github.com/OneKeyHQ/tradingview-charting-library/pull/197

The native settings-overlay fix was merged separately in app PR #12415 and chart PR #196, so it is intentionally not duplicated in this PR's final diff.

## Root Cause
The app planned non-native history slices for up to 2,000 points, while the market endpoint returned only roughly the latest 299 points for wide requests. TradingView received that truncated tail with `noData: false`, treated the larger interval as loaded, and left a permanent gap containing the selected date range.

## Design Decisions
- Limit non-native history slices to 200 points, leaving margin below the observed backend cap.
- Overlap adjacent requests by one interval because endpoint boundaries are exclusive, then deduplicate by timestamp after all slices finish.
- Clip the merged result to TradingView's `[from, to)` contract so the upper-bound candle is not duplicated.
- Keep request planning/merge in the app `main` JS runtime. HTTP calls continue through the independent `bg` runtime; data crosses runtimes as serialized JS copies and no shared native K-line state is introduced.

## Changes Detail
- K-line slicing: configurable safe slice size and one-interval boundary overlap.
- K-line merge: exclusive upper-bound clipping and timestamp deduplication.
- Tests: backend-cap regression coverage plus slice continuity and boundary cases.

## Risk Assessment
- **Risk Level**: Medium
- **Affected Platforms**: Extension / Mobile / Desktop / Web
- **Risk Areas**: Increased concurrent history requests for very wide ranges, endpoint boundary semantics, and coordinated rollout with the chart-side pagination/resolution fix.

## Test plan
- [x] `yarn agent:check --profile commit`
- [x] Targeted Jest suites (50 tests)
- [x] Targeted ESLint, Prettier, and `git diff --check`
- [x] Verify the May date-range trace returns 790 continuous `1H` bars and the exact requested initial visible range
- [ ] Manually verify go-to-date and time-range navigation in the embedded app host

[OK-57783]: https://onekeyhq.atlassian.net/browse/OK-57783?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[OK-57784]: https://onekeyhq.atlassian.net/browse/OK-57784?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by originalix** [comment]
## Automated code review found blocking issues

@limichange

Reviewed commit `2df5defb791b`. This report always reflects the latest reviewed changes.

### Review summary

This PR changes the TradingViewV2 history-data path across the foreground/background boundary: the main runtime now splits non-native market K-line windows into approximately 200-point requests with a one-bar overlap, sends them through `backgroundApiProxy` to the bg-owned market HTTP service, and deduplicates the results into a half-open `[timeFrom, timeTo)` window. This is intended to prevent backend response caps and exclusive slice boundaries from dropping chart bars, with focused regression coverage. Package ownership and security boundaries remain unchanged, but the tighter slicing materially increases cross-runtime request fan-out while retaining fail-fast aggregation, so one failed slice can discard all successful data.

**What needs attention:** Confirm that wide `1m` and `1H` history loads use bounded concurrency and preserve successful slice responses when one request fails.

### Issues to address

- **P1 · Tighter slices create an unbounded fail-fast request burst**
  - **Impact:** For non-native tokens, the existing two-day minimum combined with `maxDataLength: 200` turns a `1m` history load into 15 cross-runtime HTTP requests, and longer ranges create more. The unchanged `Promise.all` launches every slice at once and rejects on the first failure, discarding successful responses. Most TradingViewV2 consumers do not provide a fallback, so the function returns `null` and the message handler sends no history response, leaving the chart empty or waiting.
  - **Suggested action:** Execute slices through a bounded-concurrency all-settled pool, merge successful responses, and use the fallback only when no slice succeeds.

### Validation gaps

- `git diff --check origin/x...HEAD` completed without whitespace errors.
- The targeted Jest command for both changed test files could not run because the worktree lacks the Yarn `node_modules` state file.
- No live TradingView/backend integration was available to measure request throttling, partial failures, or chart recovery across iframe and injected-script transports.

**Comment by sidmorizon** [comment]
@codex review

**Comment by sidmorizon** [comment]
@codex security review

**Comment by sidmorizon** [comment]
@cursoragent review

File: packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts [hunk 1]
```diff
@@ -401,6 +403,9 @@ export async function fetchMarketKLineDataWithSlicing({
     const isNativeToken = !tokenAddress;
     const slices = sliceKLineRequest(interval, timeFrom, timeTo, {
       isNativeToken,
+      ...(!isNativeToken
+        ? { maxDataLength: MARKET_KLINE_MAX_POINTS_PER_REQUEST }
+        : {}),
```
Comment (sidmorizon) [inline packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts:418]: **One failed slice blanks history**

Severity: `non-severe`

After `maxDataLength` multiplies slice requests, one rejection aborts `Promise.all` and discards every successful slice. The chart receives fallback data or no history.

File: packages/kit/src/components/TradingView/utils/sliceKLineRequest.ts [hunk 2]
```diff
@@ -73,7 +76,11 @@ export function sliceKLineRequest(
   const timePerSlice = Math.floor((timeTo - adjustedTimeFrom) / sliceCount);
 
   return Array.from({ length: sliceCount }, (_, index) => ({
-    from: adjustedTimeFrom + index * timePerSlice,
+    // Overlap one interval because the market K-line endpoint excludes slice boundaries.
+    from:
+      adjustedTimeFrom +
+      index * timePerSlice -
+      (index > 0 ? intervalSeconds : 0),
```
Comment (sidmorizon) [inline packages/kit/src/components/TradingView/utils/sliceKLineRequest.ts]: **Requested boundary candle remains missing**

Severity: `non-severe`

When `timeFrom` aligns with a candle, the first slice's `from` starts there despite the endpoint's exclusive lower bound. Both adjoining responses omit that candle.

File: packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts [hunk 3]
```diff
@@ -401,31 +406,53 @@ export async function fetchMarketKLineDataWithSlicing({
...
+      maxSliceCount: MARKET_KLINE_MAX_SLICE_COUNT,
```
Comment (sidmorizon) [inline packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts]: **Wide history requests return no data**

Severity: `non-severe`

Without fallback data, requests above 19,901 candles make `maxSliceCount` abort retrieval. Minute charts fail beyond roughly 14 days.

File: packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts [hunk 4]
```diff
@@ -401,6 +403,9 @@ export async function fetchMarketKLineDataWithSlicing({
     const isNativeToken = !tokenAddress;
     const slices = sliceKLineRequest(interval, timeFrom, timeTo, {
       isNativeToken,
+      ...(!isNativeToken
+        ? { maxDataLength: MARKET_KLINE_MAX_POINTS_PER_REQUEST }
```
[the remark this answers is not in the document]
Comment (limichange) [reply packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts:417]: Fixed in  and : sliced K-line requests now use bounded sliding-window concurrency, share a hard budget of 100 total backend attempts (including retries), retry failed/invalid slices within the remaining budget, and merge only when every slice is valid. Otherwise the whole request follows the existing fallback path. Regression coverage includes transient recovery, the 100-call budget, concurrency, and incomplete-history rejection.

File: packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (limichange) [reply packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts:418]: Fixed in  and : the caller now rejects plans over 100 slices before fan-out, schedules at  (8 Native / 10 elsewhere), and keeps initial requests plus retries within 100 total backend calls. The five-year/1m extreme now fails planning instead of creating ~13k requests.

[the remark this answers is not in the document]
Comment (limichange) [reply packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts:418]: Fixed in  and : all planned slice results are validated, invalid slices are retried within the global request budget, and any unresolved invalid slice sends the whole request through fallback rather than merging partial history. Covered by .

Comment (limichange) [reply packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts:418]: Fixed in : a slice rejection no longer aborts a raw . Each slice settles independently under bounded concurrency, failed slices are retried within the 100-attempt budget, and only a complete valid set is returned; an unrecoverable slice uses fallback instead of returning a gapped result.

File: packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (limichange) [reply packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts:417]: Correction with commit references: fixed by ceea1747ef and 2923cd99ad. Sliced K-line requests now use bounded sliding-window concurrency, share a hard budget of 100 total backend attempts including retries, retry failed or invalid slices within the remaining budget, and merge only when every slice is valid. Otherwise the whole request follows fallback. Regression coverage includes transient recovery, the 100-call budget, concurrency, and incomplete-history rejection.

File: packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (limichange) [reply packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts:418]: Correction with commit references: fixed by ceea1747ef and 2923cd99ad. The caller rejects plans over 100 slices before fan-out, schedules at the shared platform concurrency limit of 8 on Native and 10 elsewhere, and keeps initial requests plus retries within 100 total backend calls. The five-year one-minute extreme now fails planning instead of creating about 13,000 requests.

[the remark this answers is not in the document]
Comment (limichange) [reply packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts:418]: Correction with commit references: fixed by ceea1747ef and 2923cd99ad. All planned slice results are validated, invalid slices are retried within the global request budget, and any unresolved invalid slice sends the whole request through fallback rather than merging partial history. This is covered by the incomplete sliced-history regression test.

Comment (limichange) [reply packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts:418]: Correction with commit reference: fixed by 2923cd99ad. A slice rejection no longer aborts a raw Promise.all operation. Each slice settles independently under bounded concurrency, failed slices are retried within the 100-attempt budget, and only a complete valid set is returned. An unrecoverable slice uses fallback instead of returning a gapped result.

File: packages/kit/src/components/TradingView/utils/sliceKLineRequest.ts (hunk 2 above)
Comment (limichange) [reply packages/kit/src/components/TradingView/utils/sliceKLineRequest.ts]: Fixed in 170ffdbeec. The first K-line request now pads the exclusive lower bound by one second, so a candle exactly at TradingView timeFrom is returned. Internal boundaries remain covered by the existing one-interval overlap, and the change does not add slices or consume extra request budget. Added an exclusive-endpoint regression test.

File: packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts (hunk 3 above)
Comment (limichange) [reply packages/kit/src/components/TradingView/utils/fetchMarketKLineData.ts]: Accepted as the intentional safety boundary. A five-year one-minute range would plan roughly 13,207 backend calls after strict overlap, so plans above 100 slices must fail closed and use the existing fallback path. The companion chart change is responsible for requesting a coarser resolution or smaller pages. Returning partial history or allowing the fan-out would reintroduce the original gap and resource-exhaustion risks. The 100-slice and 100-total-attempt limits are covered by regression tests.

**Comment by limichange** [comment]
Ablation study completed against the final 59-test oracle in an isolated detached worktree. Only the two production blobs were swapped; tests stayed fixed.

| Variant | Result | Lost behavior |
| --- | ---: | --- |
| Final implementation | 59/59 | None |
| Remove exclusive lower-bound padding | 57/59 | Aligned from candle is omitted |
| Also remove bounded retry budget | 55/59 | Transient slice recovery and 100-attempt budget fail |
| Original PR before hardening | 49/59 | Strict slice sizing, 100-slice rejection, completeness, retry, bounded concurrency, and lower boundary fail |
| Remove all PR production changes | 43/59 | Wide-range continuity also fails |
| Final orchestration + base slicer | 49/59 | Planner overlap, safe window size, and planner cap fail |
| Base orchestration + final slicer | 52/59 | Caller limit, completeness, retry, concurrency, and lower boundary fail |

Concrete baseline impact: for the 1,531-hour capped-response fixture, the base implementation returns only 299 bars and misses 1,232 bars (80.5%). At the extreme, five years at 1m would require about 13,207 strict-overlap slices without the cap. The final implementation rejects plans above 100 slices, limits active requests to the shared platform limit, and keeps initial calls plus retries at no more than 100 total attempts.

Conclusion: planner and orchestration are independently necessary; the retry and lower-bound layers each protect a distinct failure mode. No experimental changes remain in the working tree.

**Comment by sidmorizon** [comment]
@cursoragent review

**Comment by sidmorizon** [comment]
@cursoragent review

**Review by originalix — APPROVED** [review]
